canonical/maas-ui
# feat(machines): update DHCP and discovery forms to use the new API #1116

## Done

- update DHCP add form to use list box
- Update discovery form to use the new API and machine list box
- keep previous data to prevent flashes of no data in between data requests
- add e2e test for dhcp snippets
- add `SelectButton` component
- add OutsideClickHandler component

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to `MAAS/r/settings/dhcp`
- Select Type: Machine and pick a machine from the list 
- Go to network discovery
- Expand the dropdown on the right hand side and select "Add discovery..."
- Select a machine from the list in the Parent field and
- Ensure the device.create websocket message has been sent with the correct payload (including the id of a machine which was selected as a parent)
- You can look at websocket messages using developer tools
  - go to network tab -> filter by WS 
  - look for an entry starting with `ws?csrftoken=` 
- You can verify the parent id is valid by visiting the machine details page with a given id and verifying it's the one that was requested in the form (e.g. `/MAAS/r/machine/[replace-this-with-your-selected-parent-id]`)


## Fixes

Fixes: https://github.com/canonical/app-tribe/issues/1116
Fixes: https://github.com/canonical/app-tribe/issues/1112
Fixes: https://github.com/canonical/app-tribe/issues/1340

## Launchpad issue

Related Launchpad maas issue in the form `lp#number`.

## Backports

In general, please propose fixes against _main_ rather than release branches (e.g. 2.7), unless the fix is only applicable for that specific release. Please apply backport labels to the PR (e.g. "Backport 2.7") for the appropriate releases to target.

Only bug and security fixes should be backported, new features should only land in main.

## Screenshots


![Kapture 2022-09-07 at 15 55 02](https://user-images.githubusercontent.com/7452681/188896348-8b71ee3d-9d84-4ae9-8b2d-e26042bb0c7c.gif)
